tests/HelperTest.php

```diff
@@ -0,0 +1,438 @@
+<?php
+declare(strict_types=1);
+
+namespace Emailcenter\MaxemailApi;
+
+use GuzzleHttp\Client as GuzzleClient;
+use GuzzleHttp\Exception\ClientException;
+use GuzzleHttp\Handler\MockHandler;
+use GuzzleHttp\HandlerStack;
+use GuzzleHttp\Middleware as GuzzleMiddleware;
+use GuzzleHttp\Psr7\Request;
+use GuzzleHttp\Psr7\Response;
+use phpmock\phpunit\PHPMock;
+use PHPUnit\Framework\TestCase;
+
+class HelperTest extends TestCase
+{
+    use PHPMock;
+
+    /**
+     * @var Client|\PHPUnit_Framework_MockObject_MockObject
+     */
+    private $apiClientMock;
+
+    /**
+     * @var GuzzleClient|\PHPUnit_Framework_MockObject_MockObject
+     */
+    private $httpClientMock;
+
+    /**
+     * @var MockHandler
+     */
+    private $mockHandler;
+
+    private $clientHistory = [];
+
+    protected function setUp()
+    {
+        $this->apiClientMock = $this->createMock(Client::class);
+        $this->httpClientMock = $this->createMock(GuzzleClient::class);
+
+        $this->mockHandler = new MockHandler();
+
+        // This allows fopen() to be overloaded after it's first used normally in other tests
+        $this->defineFunctionMock(__NAMESPACE__, 'fopen');
+        $this->defineFunctionMock(__NAMESPACE__, 'fwrite');
+        $this->defineFunctionMock(__NAMESPACE__, 'fclose');
+        $this->defineFunctionMock(__NAMESPACE__, 'unlink');
+    }
+
+    public function testUpload()
+    {
+        $sampleFile = __DIR__ . '/__files/sample-file.csv';
+        $key = 'abc123def456';
+
+        /** @var Service|\PHPUnit_Framework_MockObject_MockObject $fileUploadService */
+        $fileUploadService = $this->createMock(Service::class);
+
+        // The API client will be used to initialise the upload
+        $this->apiClientMock->expects($this->once())
+            ->method('__get')
+            ->with('file_upload')
+            ->willReturn($fileUploadService);
+
+        $fileUploadService->expects($this->once())
+            ->method('__call')
+            ->with('initialise', [])
+            ->willReturn((object)['key' => $key]);
+
+        // HTTP Client is used to perform multipart upload
+        // Can't use MockHandler and History middleware,
+        // as once the request has been handled the file stream cannot be read again
+        $checkMultipart = function($options) use ($key) {
```

Comment:
Agreed. I wanted to use the MockHandler and history middleware to check the request body contained the expected multipart content (so that the bit that Maxemail receives in valid).

Doesn't seem possible though, as the history middleware simply keeps the instance of the request, but as that request has already read the file stream, it causes a fatal error when trying to read it again.